File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/services/CheckerSchedulerService.java
package io.github.sanyavertolet.edukate.backend.services;

import io.github.sanyavertolet.edukate.backend.entities.Submission;
import io.github.sanyavertolet.edukate.messaging.RabbitTopology;
import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

@Slf4j
@Component
@AllArgsConstructor
public class CheckerSchedulerService {
    private final RabbitTemplate rabbitTemplate;
    private final SubmissionService submissionService;

    public Mono<Void> scheduleCheck(@NonNull Submission submission) {
        return submissionService.prepareContext(submission)
                .publishOn(Schedulers.boundedElastic())
                .flatMap(ctx -> Mono.fromRunnable(() ->
                        rabbitTemplate.convertAndSend(RabbitTopology.EXCHANGE, RabbitTopology.Rk.SCHEDULE, ctx)
                ));
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/services/CheckerSchedulerService.kt
package io.github.sanyavertolet.edukate.backend.services

import io.github.sanyavertolet.edukate.backend.entities.CheckResult
import io.github.sanyavertolet.edukate.backend.entities.Submission
import io.github.sanyavertolet.edukate.backend.mappers.SubmissionMapper
import io.github.sanyavertolet.edukate.backend.repositories.CheckResultRepository
import io.github.sanyavertolet.edukate.messaging.RabbitTopology
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers

@Component
class CheckerSchedulerService(
    private val rabbitTemplate: RabbitTemplate,
    private val submissionMapper: SubmissionMapper,
    private val checkResultRepository: CheckResultRepository,
) {
    fun scheduleCheck(submission: Submission): Mono<Void> {
        val submissionId = requireNotNull(submission.id) { "Submission ID must not be null" }
        return checkResultRepository.save(CheckResult.stub(submissionId)).flatMap { stub ->
            val checkResultId = requireNotNull(stub.id) { "CheckResult ID must not be null after save" }
            submissionMapper.prepareContext(submission, checkResultId).publishOn(Schedulers.boundedElastic()).flatMap { ctx
                ->
                Mono.fromRunnable { rabbitTemplate.convertAndSend(RabbitTopology.EXCHANGE, RabbitTopology.Rk.SCHEDULE, ctx) }
            }
        }
    }
}
